### DISCUSSION TITLE
LoRaWAN: how to limit TX power?

### DISCUSSION THREAD

**Discussion by tve** [jgromes/RadioLib#1287] [Q&A]
I'm using an stm32wle5 and need to limit the TX power to 10dBm due to battery constraints, what is the best way to achieve that? Calling `node.setTxPower(10)` doesn't affect OTAA as far as I can tell. I believe I need to set `node.txPowerMax` which is protected. Thoughts?

**Comment by StevenCellist** [comment]
There is no clean way of handling this within the LoRaWAN stack. The cleanest way I can think of is restricting the module driver to 10dBm max ([here](https://github.com/jgromes/RadioLib/blob/fada24c1c4e4cc65300309604506a209fbe652e3/src/modules/SX126x/STM32WLx.cpp#L62) would be the right place I think).
By the way, out of curiosity, are you running through STM32duino or through STCube?

**Reply by HeadBoffin** [reply]
And adjust the ADR processing so that it doesn't get hijacked - or turn it off.

**Reply by StevenCellist** [reply]
Actually, that isn't even necessary at all - the ADR specification requires that a node is able to "operate at or below" the requested Tx power. As such the stack only sets the highest possible value out of (LoRaWAN, module), which might mean the module is the limiting factor. But it perfectly falls within spec :)

**Reply by HeadBoffin** [reply]
I meant that when an ADR request comes down the line, you'll need to restore the power settings - as the algorithm sets max power until it gets to SF7 then it start telling it reduce power if link budget allows.

**Reply by StevenCellist** [reply]
That is true for the case where you just use `node.setTxPower(10)`, but if you modify the underlying module driver and limit its range to 10dBm, the stack will (should) never exceed this range.
